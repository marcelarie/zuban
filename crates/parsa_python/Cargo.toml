[package]
name = "parsa_python"
version.workspace = true
edition.workspace = true
rust-version.workspace = true
publish = false
homepage.workspace = true
authors.workspace = true

[dependencies]
parsa.workspace = true
lazy_static.workspace = true
regex.workspace = true

[dev-dependencies]
insta.workspace = true
test_utils.workspace = true
