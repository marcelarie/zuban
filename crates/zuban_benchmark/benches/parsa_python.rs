use criterion::{criterion_group, criterion_main, Criterion};

use parsa_python_cst::Tree;

fn parse_file(copies: usize) {
    let some_code = utils::dedent(
        r#"
        import os
        def x(a, b, c):
            try:
                return a
            except Exception:
                return [1, 2, 3, 4, 5, 6]
        class C:
            x: int
            y = 1 # type: int
            z: dict[int, list[str]] = {1: ["", r''], 2: [f""" {1} """]}
            def f(self, x: list[str], y: int): ...
    "#,
    );
    assert_eq!(some_code.len(), 262);
    Tree::parse(some_code.repeat(copies).into());
}

fn bench_parser(c: &mut Criterion) {
    c.bench_function("parse big file", |b| b.iter(|| parse_file(1000)));
    c.bench_function("parse small file", |b| b.iter(|| parse_file(1)));
}

// Register the benchmarks
criterion_group!(benches, bench_parser);
criterion_main!(benches);
