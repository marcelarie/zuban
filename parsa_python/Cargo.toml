[package]
name = "parsa_python"
version = "0.1.0"
edition.workspace = true
rust-version.workspace = true
homepage.workspace = true
authors.workspace = true

[dependencies]
parsa.workspace = true
lazy_static.workspace = true
regex.workspace = true

[dev-dependencies]
insta.workspace = true
