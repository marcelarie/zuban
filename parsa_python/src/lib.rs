use parsa::{create_parser,create_node,create_token};


create_token!(PythonToken, enum TokenType, String, Number, Endmarker);
create_node!(PythonNode, enum NodeType, file_input, foo);

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn it_works() {
        assert_eq!(2 + 2, 4);
        parse_python("bar");
        dbg!(TokenType::get_map());
        dbg!(NodeType::get_map());
        return
    }
}

struct PythonTokenizer {
}

impl PythonTokenizer {
    fn new_tok(&self) -> PythonToken {
        PythonToken {
            start: 0,
            length: 0,
            type_: TokenType::Endmarker,
        }
    }
}

impl parsa::Tokenizer<PythonToken> for PythonTokenizer {
    fn new(_code: &str) -> Self {
        Self {}
    }

    fn yield_next(&self) -> PythonToken {
        self.new_tok()
    }

}

create_parser!(parse_python, PythonTree, PythonNode,
               PythonToken, PythonTokenizer, TokenType, NodeType);
//create_parser!(parse_pythonx, PythonTreex, PythonNodex,
//               PythonToken, PythonTokenizer, TokenType, NodeType);
