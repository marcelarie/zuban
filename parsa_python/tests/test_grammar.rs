extern crate insta;

use parsa_python::*;

fn dedent(s: &'static str) -> String {
    let mut indent = usize::MAX;
    let lines: &Vec<_> = &s.split('\n').collect();
    for line in lines {
        if !line.trim().is_empty() {
            indent = std::cmp::min(indent, line.len() - line.trim_start().len());
        }
    }
    if indent == usize::MAX {
        return s.to_string();
    }
    let new_lines: Vec<_> = lines
        .iter()
        .map(|line| {
            if line.len() >= indent {
                &line[indent..]
            } else {
                line
            }
        })
        .collect();
    new_lines.join("\n")
}

fn tree_to_string(tree: PyTree) -> String {
    fn recurse(code: &mut String, node: &PyNode, depth: usize) {
        *code += &" ".repeat(depth);
        *code += &format!(
            "{}: {}-{}{}\n",
            node.type_str(),
            node.start(),
            node.end(),
            if node.is_leaf() {
                format!(" {:?}", node.as_code())
            } else {
                "".to_string()
            }
        );
        for c in node.iter_children() {
            assert_eq!(node.index, c.parent().unwrap().index);
            recurse(code, &c, depth + 1);
        }
    }

    let root = tree.root_node();
    assert!(root.parent().is_none());
    let mut code = String::new();
    recurse(&mut code, &root, 0);
    code
}

macro_rules! parametrize_snapshots {
    ($($name:ident : $input:expr;)*) => {$(
        #[test]
        fn $name() {
            let tree = parse($input.into());
            insta::assert_snapshot!(stringify!($name), tree_to_string(tree));
        }
    )*}
}

parametrize_snapshots!(
    empty: "".to_owned();
    very_simple: "1".to_owned();
    json: "{foo: 1}\n".to_owned();
    simple: dedent("
        a(bar:=1, foo=1);
        if a:
         1
        else: 2
        ");
    assignments_simple: dedent("
        a = 1
        a, b = 2
        a, b, = 2
        (a, b) = 2
        ((a, b)) = 2
        a = b = 3
        a = (b, c) = 3
        a = (b, c,) = 3
        a = b, c, = 3
        ");
    assignments_star: dedent("
        *foo = *bar = *baz
        i, *foo = ((j, [*bar, k])) = *baz
        *foo, *bar = yield baz
        ");
    assignments_aug: dedent("
        foo += 3
        foo.bar |= 3
        foo, bar += 3
        ");
    assignments_setitem: dedent("
        foo[0] += 3
        foo[0] = a = 3
        foo[a, b] = c
        foo[a::, b:f, c:=3] = c
        ");
    assignments_annotation: dedent("
        foo: bar
        foo: int = 3
        foo.bar: baz
        foo: bar = a = 4
        ");
    assignments_with_call: dedent("
        foo().bar = 1
        foo(1,).bar[1] = 1
        foo(a for a in b).bar[1] = 1
        ");
    for_stmt: dedent("
        for x in [1,2]:
            ...
        for *b in *c: ...
        for c.x, (*b, d.y[0]) in i, *j, k: ...
        ");
    comprehension_simple: dedent("
        [1 for a in b]
        [a for *a, c in b if c if d]
        foo(a for (*a, (*c, b),), in d)
        ");
    comprehension_fails: dedent("
        # fails
        [1 for a in *b]
        [1 for a in b if a else c]
        ");
    del_stmt: dedent("
        del foo
        del foo,
        del foo, bar.baz
        del foo[1]
        ");
    cls: dedent("
        class Foo(object):
            def __init__(self, /, f, *, g):
                pass

            @property
            def prop(self):
                pass
        ");
    import_names: dedent("
        import foo
        import foo.bar as baz
        import foo, bar, baz
        import foo.bar.baz
        ");
    import_name_fails: dedent("
        import foo,
        import foo, bar,
        ");
    import_from_normal: dedent("
        from foo import bar
        from foo import bar as baz
        from foo.bar.baz import *
        ");
    import_from_dotted: dedent("
        from . import (foo as baz, bar)
        from . import (foo, bar,)
        from ..foo import bar as baz
        from ... import *
        ");
    import_from_fails: dedent("
        from . import foo,
        from . import foo, bar,
        ");
    calls: dedent("
        foo(bar)
        foo(bar_foo=3)
        foo(bar_baz:=2)
        foo(**kwargs)
        ");
    nested_call: dedent("
        foo(1)()
        ");
    complex_calls: dedent("
        foo(bar, 3)
        foo(bar, *foo)
        foo(bar, *foo, name=3)
        foo(bar, *foo, name=3, **kwargs)
        foo(bar, *foo, name=3, **kwargs, b=3)
        foo(**kwargs, b=3, **foo)
        ");
    operators: dedent("
        f | d | c + 5 & 2 * 3 * 4 ** 5 ** 6
        ");
    simple_error_recovery: dedent("
        a + 3 +
        b = 3
        ");
    terminal_error_recovery: dedent("
        ?
        ");
    terminal_and_nonterminal_error_recovery: dedent("
        foo
        (1) + ? + c
        bar
        ");
    keyword_recovery: dedent("
        foo
        def
        bar
        else:
        baz
        ");
    match_simple: dedent("
        match foo:
            case ['a']: ...
            case _:
                pass
        ");
    soft_keyword_underscore: dedent("
        match _:
            case ['a']: _(_)
            case _:
                _
        match = 3
        ");
    match_underscore_with_as: dedent("
        match foo:
            case bar as _: pass
        ");
    match_class_pattern: dedent("
        match a:
            case A():
                pass
            case B(1, 2):
                pass
            case B(1, b=2):
                pass
            case B(a=1, b=2):
                pass
        ");
    dict_literal1: dedent("
        {1: 2}
        {**foo}
        {**foo, 1: 2}
        {**foo, **bar}
        ");
    function_simple: dedent("
        def foo(): ...
        def foo(a: 3): ...
        def foo(a, b,): ...
        def foo(a: f = 3, b=4): ...
        def foo(a=3, b=4): ...
        ");
    function_args: dedent("
        def foo(a, *, b=3): ...
        def foo(a=3, *, c): ...
        def foo(a, *, x, **kwargs): ...
        def foo(a, *args: int, **fooo: str,): ...
        def foo(a, /, b: i, c: i=3, **kwargs): ...
        def foo(*args): ...
        ");
    lambda_simple: dedent("
        lambda: ...
        lambda a: ...
        lambda a, b,: ...
        lambda a=3, b=4: ...
        ");
    lambda_slash: dedent("
        lambda a, /, b=3: ...
        lambda a=3, /,: ...
        lambda a=3, /: ...
        lambda a=3, /, *args, c=3: ...
        lambda a=3, /, c=3: ...
        lambda a, /, b, c=3, **kwargs: ...
        lambda a, a=3, /, c=3, *, d: ...
        ");
    lambda_args: dedent("
        lambda a, *, b=3: ...
        lambda a=3, *, c: ...
        lambda a, *, x, **kwargs: ...
        lambda a, *args, **fooo ,: ...
        lambda *args: ...
        ");
    lambda_kwargs: dedent("
        lambda **kwargs: ...
        lambda a=3, **kwargs: ...
        lambda a, **kwargs: ...
        lambda a, /, b, *, x, **kwargs: ...
        ");
    lambda_fails: dedent("
        lambda a, *, **kwargs: ...
        lambda a=3, b: ...
        lambda **a, b: ...
        lambda a=3, *,: ...
        lambda ,: ...
        ");
    lambda_fail_with_slash: dedent("
        lambda a=3, /, b: ...
        lambda a=3, /, b=3, c: ...
        lambda a, /, b=3, c: ...
        lambda z, a=3, d, /, b: ...
        ");
    with_items_simple: dedent("
        with foo as bar: ...
        with foo as bar, baz: ...
        ");
    with_items_parentheses: dedent("
        with (foo as bar): ...
        with (foo as bar, baz): ...
        with ((foo)): ...
        with (i for i in 42): ...
        with (yield y): ...
        ");
    string_literals_in_strings: dedent(r#"
        '"""'
        "'''"
        ""
        ''
        "#);
    bytes: dedent(r#"
        b'asdf\n\''
        br'\asdf'
        Br'asdf"""'
        bR''
        BR'\''
        "#);
    exception_group: dedent(r#"
        try:
            ...
        except* SpamError:
            ...
        except* FooError as e:
            ...
        except* (BarError, BazError) as e:
            ...
        "#);
    invalid_syntax: dedent(r#"
        def f():
            (x,) += 1
            z
        "#);
);
