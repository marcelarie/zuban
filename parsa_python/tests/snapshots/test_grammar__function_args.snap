---
source: parsa_python/tests/test_grammar.rs
expression: tree_to_string(tree)
---
file: 0-186
 stmt: 1-25
  compound_stmt: 1-25
   function_def: 1-25
    Keyword: 1-4 "def"
    name_definition: 5-8
     Name: 5-8 "foo"
    Keyword: 8-9 "("
    parameters: 9-18
     param_no_default: 9-10
      name_definition: 9-10
       Name: 9-10 "a"
     Keyword: 10-11 ","
     star_etc: 12-18
      Keyword: 12-13 "*"
      Keyword: 13-14 ","
      param_maybe_default: 15-18
       name_definition: 15-16
        Name: 15-16 "b"
       Keyword: 16-17 "="
       expression: 17-18
        bitwise_or: 17-18
         Number: 17-18 "3"
    Keyword: 18-19 ")"
    Keyword: 19-20 ":"
    block: 21-25
     simple_stmts: 21-25
      simple_stmt: 21-24
       star_expressions: 21-24
        expression: 21-24
         bitwise_or: 21-24
          Keyword: 21-24 "..."
      Newline: 24-25 "\n"
 stmt: 25-49
  compound_stmt: 25-49
   function_def: 25-49
    Keyword: 25-28 "def"
    name_definition: 29-32
     Name: 29-32 "foo"
    Keyword: 32-33 "("
    parameters: 33-42
     param_with_default: 33-36
      name_definition: 33-34
       Name: 33-34 "a"
      Keyword: 34-35 "="
      expression: 35-36
       bitwise_or: 35-36
        Number: 35-36 "3"
     Keyword: 36-37 ","
     star_etc: 38-42
      Keyword: 38-39 "*"
      Keyword: 39-40 ","
      param_maybe_default: 41-42
       name_definition: 41-42
        Name: 41-42 "c"
    Keyword: 42-43 ")"
    Keyword: 43-44 ":"
    block: 45-49
     simple_stmts: 45-49
      simple_stmt: 45-48
       star_expressions: 45-48
        expression: 45-48
         bitwise_or: 45-48
          Keyword: 45-48 "..."
      Newline: 48-49 "\n"
 stmt: 49-81
  compound_stmt: 49-81
   function_def: 49-81
    Keyword: 49-52 "def"
    name_definition: 53-56
     Name: 53-56 "foo"
    Keyword: 56-57 "("
    parameters: 57-74
     param_no_default: 57-58
      name_definition: 57-58
       Name: 57-58 "a"
     Keyword: 58-59 ","
     star_etc: 60-74
      Keyword: 60-61 "*"
      Keyword: 61-62 ","
      param_maybe_default: 63-64
       name_definition: 63-64
        Name: 63-64 "x"
      Keyword: 64-65 ","
      double_starred_param: 66-74
       Keyword: 66-68 "**"
       name_definition: 68-74
        Name: 68-74 "kwargs"
    Keyword: 74-75 ")"
    Keyword: 75-76 ":"
    block: 77-81
     simple_stmts: 77-81
      simple_stmt: 77-80
       star_expressions: 77-80
        expression: 77-80
         bitwise_or: 77-80
          Keyword: 77-80 "..."
      Newline: 80-81 "\n"
 stmt: 81-123
  compound_stmt: 81-123
   function_def: 81-123
    Keyword: 81-84 "def"
    name_definition: 85-88
     Name: 85-88 "foo"
    Keyword: 88-89 "("
    parameters: 89-116
     param_no_default: 89-90
      name_definition: 89-90
       Name: 89-90 "a"
     Keyword: 90-91 ","
     star_etc: 92-116
      starred_param: 92-102
       Keyword: 92-93 "*"
       name_definition: 93-97
        Name: 93-97 "args"
       annotation: 97-102
        Keyword: 97-98 ":"
        expression: 99-102
         bitwise_or: 99-102
          Name: 99-102 "int"
      Keyword: 102-103 ","
      double_starred_param: 104-115
       Keyword: 104-106 "**"
       name_definition: 106-110
        Name: 106-110 "fooo"
       annotation: 110-115
        Keyword: 110-111 ":"
        expression: 112-115
         bitwise_or: 112-115
          Name: 112-115 "str"
      Keyword: 115-116 ","
    Keyword: 116-117 ")"
    Keyword: 117-118 ":"
    block: 119-123
     simple_stmts: 119-123
      simple_stmt: 119-122
       star_expressions: 119-122
        expression: 119-122
         bitwise_or: 119-122
          Keyword: 119-122 "..."
      Newline: 122-123 "\n"
 stmt: 123-166
  compound_stmt: 123-166
   function_def: 123-166
    Keyword: 123-126 "def"
    name_definition: 127-130
     Name: 127-130 "foo"
    Keyword: 130-131 "("
    parameters: 131-159
     param_no_default: 131-132
      name_definition: 131-132
       Name: 131-132 "a"
     Keyword: 132-133 ","
     Keyword: 134-135 "/"
     Keyword: 135-136 ","
     param_no_default: 137-141
      name_definition: 137-138
       Name: 137-138 "b"
      annotation: 138-141
       Keyword: 138-139 ":"
       expression: 140-141
        bitwise_or: 140-141
         Name: 140-141 "i"
     Keyword: 141-142 ","
     param_with_default: 143-149
      name_definition: 143-144
       Name: 143-144 "c"
      annotation: 144-147
       Keyword: 144-145 ":"
       expression: 146-147
        bitwise_or: 146-147
         Name: 146-147 "i"
      Keyword: 147-148 "="
      expression: 148-149
       bitwise_or: 148-149
        Number: 148-149 "3"
     Keyword: 149-150 ","
     star_etc: 151-159
      double_starred_param: 151-159
       Keyword: 151-153 "**"
       name_definition: 153-159
        Name: 153-159 "kwargs"
    Keyword: 159-160 ")"
    Keyword: 160-161 ":"
    block: 162-166
     simple_stmts: 162-166
      simple_stmt: 162-165
       star_expressions: 162-165
        expression: 162-165
         bitwise_or: 162-165
          Keyword: 162-165 "..."
      Newline: 165-166 "\n"
 stmt: 166-186
  compound_stmt: 166-186
   function_def: 166-186
    Keyword: 166-169 "def"
    name_definition: 170-173
     Name: 170-173 "foo"
    Keyword: 173-174 "("
    parameters: 174-179
     star_etc: 174-179
      starred_param: 174-179
       Keyword: 174-175 "*"
       name_definition: 175-179
        Name: 175-179 "args"
    Keyword: 179-180 ")"
    Keyword: 180-181 ":"
    block: 182-186
     simple_stmts: 182-186
      simple_stmt: 182-185
       star_expressions: 182-185
        expression: 182-185
         bitwise_or: 182-185
          Keyword: 182-185 "..."
      Newline: 185-186 "\n"
 Endmarker: 186-186 ""

